### PR TITLE
Change WiFi Config

### PR DESCRIPTION
**Pull request by jeremypoulter**
Changes to allow the WiFi config to be changed while in client mode. Click on the WiFi network details to open the WiFi scan.

This also handles setting the WiFi after a failure to connect to the configured WiFi network falls back to AP mode (#118)

### CODE REVIEW

File: src/wifi.cpp [hunk 1]
```diff
@@ -282,7 +282,7 @@ wifi_loop() {
   }
 
   // Remain in AP mode for 5 Minutes before resetting
-  if(isApOnly && client_retry && millis() > client_retry_time) {
+  if(isApOnly && 0 == apClients && client_retry && millis() > client_retry_time) {
```
Comment (glynhudson) [inline src/wifi.cpp]: Nice! This is very clever

**Comment by glynhudson** [comment]
This is great, thanks a lot for taking a look at the WiFi config issue. It's one of the last remaining potentially big issues with the current FW. Sorry for the delay testing this, I just got home this week and haven't had access to an openevse to test until now. 

Switching between WiFi networks while in client mode works great although it took me a while to figure our where to click to view the drop down list of WiFi network scan. Maybe a button "Scan WiFi Networks" or "Change WiFi Network" would be more intuitive? 

openevse - edited


I switching WiFi network to a WiFi network created by my phone hotspot. This worked perfectly. I then switched off my phone hotspot to simulate a WiFi network going down. The ESP dropped into AP mode as expected. I was able to connect to the ESP AP and select my home WiFi network. However when I tired to connect to my home network the ESP tried to connect then dropped back into AP mode. A manual ESP reset was required to make the ESP connect to the home network. See video demo:

*currently uploading: I'll update with URL in 20min!*

**Comment by glynhudson** [comment]
Hi @jeremypoulter, Have you been able to replicate the issue I demonstrated above? Would you like me to test on another WiFi network if its's working ok for you? I

**Comment by jeremypoulter** [comment]
Yeah, can reproduce the issue, having a look at it now.

**Comment by jeremypoulter** [comment]
... and fixed

**Comment by glynhudson** [comment]
> ... and fixed

Great! Where is the fix? I can't see any further commits on this PR. The `change_wifi_config` branch was last updated 14 days ago. Let me know and I'll try and test. Thanks a lot.

**Comment by jeremypoulter** [comment]
Hum.... That sucks must have forgotten to push... As soon as Virgin get
their act together and fix my internet connection I will push.

**Comment by jeremypoulter** [comment]
All updated now, give it a try

**Comment by glynhudson** [comment]
Wow, works a treat. Absolutely seamless! 💯 #118 is fixed 👍 

Thanks so much, I can fully imagine the amount of work that's gone into this. I knew this was going to be a tricky bug to fix. I'm so glad you've nailed it. Handling WiFi connection/re-connection is a pain but super important to get working reliably. A number of users had reported getting locked out of their unit after WiFi SSID change or inputting the incorrect password. This should be fixed now.   

Thanks again.

**Pull request merged**